modules/frontend/retry.go

```diff
@@ -42,25 +50,23 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 
 		resp, err := r.next.Do(req)
 
-		triesLeft--
-		if triesLeft == 0 {
-			return resp, err
-		}
-
+		// do not retry if no error and response is not HTTP 5xx
 		if err == nil && resp.StatusCode/100 != 5 {
 			return resp, nil
 		}
 
-		statusCode := 0
-		if resp != nil {
-			statusCode = resp.StatusCode
+		// do not retry if GRPC error contains response that is not HTTP 5xx
+		httpResp, ok := httpgrpc.HTTPResponseFromError(err)
+		if ok && httpResp.Code/100 != 5 {
```

Comment:
Researching on which http status codes are retry-able, it seems maybe we shouldn't retry 501 (Not Implemented), but should also retry 429 (Too Many Requests).  What do you think?

Reply:
429 is intended as a kind of rate-limiting I believe, if you get this you should send less requests. So I don't want to retry this to avoid piling on even more.

If we can ensure retries are sent to a different querier, I agree it would be good to retry this as well. #856

Reply:
I think it's fine to ignore 501 since this shouldn't happen within Tempo. If we hit an endpoint that is not enabled (for instance with search) Tempo returns 404 not found.

Reply:
Yeah grpc returns 429 errors: I've seen the queriers return 429 when Tempo was getting a lot of request.

Reply:
Oh I just realized that retrying on 429 prompted this fix in #844.  Yep agree that might be good to explore later with better logic: a different querier or a backoff algorithm.